src/wallet/wallet.cpp

```diff
@@ -3343,23 +3345,27 @@ CWalletKey::CWalletKey(int64_t nExpires)
     nTimeExpires = nExpires;
 }
 
-int CMerkleTx::SetMerkleBranch(const CBlock& block)
+int CMerkleTx::SetMerkleBranch(const CBlock& block, int txIdx)
 {
     AssertLockHeld(cs_main);
     CBlock blockTmp;
 
     // Update the tx's hashBlock
     hashBlock = block.GetHash();
 
-    // Locate the transaction
-    for (nIndex = 0; nIndex < (int)block.vtx.size(); nIndex++)
-        if (block.vtx[nIndex] == *(CTransaction*)this)
-            break;
-    if (nIndex == (int)block.vtx.size())
+    if (txIdx != -1)
```

Comment:
@gandrewstone shouldn't the index always be greater than zero?  Here we're allowing a possibility that it could be less than -1

Reply:
-1 is a special value meaning "caller doesn't know the index".  I will fix to DbgAssert if indexes are sent that are < -1

Reply:
ok done